### PR TITLE
Roll versions forward

### PR DESCRIPTION
**Pull request by jlogan03**
* Move sundials version to latest (6.5.1)
* Move rust edition to 2021
* Move bindgen to latest
* Update simple_ode test to work with new version of sundials
* Roll package version to 0.3.0, reflecting breaking changes introduced by sundials

### CODE REVIEW

**Comment by jlogan03** [comment]
@Chris00 thoughts on this?

File: Cargo.toml [hunk 1]
```diff
@@ -37,4 +37,4 @@ nvecpthreads = []
 
 [package.metadata.docs.rs]
 
-features = ["build_libraries", "arkode", "cvode", "cvodes", "ida", "idas", "kinsol", "nvecopenmp", "nvecpthreads"]
\ No newline at end of file
+features = ["static_libraries", "build_libraries", "arkode", "cvode", "cvodes", "ida", "idas", "kinsol", "nvecopenmp", "nvecpthreads"]
```
Comment (Chris00) [inline Cargo.toml:44]: May you explain the new `static_libraries` feature?

**Comment by Chris00** [comment]
Shouldn't the vendored version be updated as well and be used if it is detected that the system libraries are too old (with a message to the user)?

File: Cargo.toml (hunk 1 above)
Comment (jlogan03) [reply Cargo.toml:44]: Ah, I just noticed that in a previous version, the static_libraries feature had already been added above, but was not added explicitly to the feature list. I can remove it if it was omitted there for a reason

**Comment by jlogan03** [comment]
> Shouldn't the vendored version be updated as well and be used if it is detected that the system libraries are too old (with a message to the user)?

The vendor submodule is updated to the commit corresponding to the latest release. I don't plan to change the linking logic on this PR.

**Comment by jlogan03** [comment]
I see what you mean about needing to select the sundials version during build - working on a solution for that on a different branch and I'll bring it into this one once I have something reasonable

**Comment by jlogan03** [comment]
Added some logic to the build.rs to figure out what version of SUNDIALS we built bindings for, and set a config flag based on the version to allow testing against either older or newer versions. This may drive a small amount of maintenance in the long run, but avoids making the crate specific to a given version of SUNDIALS or implementing platform-specific fallback logic

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -55,22 +67,22 @@ jobs:
...
-      if: startsWith(matrix.os, 'ubuntu') && ! contains(matrix.build, 'vendor')
-      run: |
-        sudo apt-get update && sudo apt-get -y install libsundials-dev
-    - name: Install Rust (rustup)
-      run: rustup update ${{ matrix.rust }} --no-self-update && rustup default ${{ matrix.rust }}
-      shell: bash
-    - run: rustup target add ${{ matrix.target }}
-    - run: cargo build
-    - run: cargo test
+      - uses: actions/checkout@master
```
Comment (Chris00) [inline .github/workflows/rust.yml]: Why this whitespace change?

File: Cargo.toml [hunk 3]
```diff
@@ -16,12 +16,17 @@ exclude = ["**/*.pdf","vendor/examples"]
...
+
+[build-dependencies.tree-sitter-rust]
+git = "https://github.com/tree-sitter/tree-sitter-rust"
 
 [target.'cfg(target_env = "msvc")'.build-dependencies]
 vcpkg = "0.2.15"
 
 [features]
 default = ["arkode", "cvode", "ida", "kinsol"]
+
```
Comment (Chris00) [inline Cargo.toml]: Why this extra line?

File: build.rs [hunk 4]
```diff
@@ -19,13 +18,15 @@ impl bindgen::callbacks::ParseCallbacks for ParseSignedConstants {
...
 // Get environment variable from string
 fn get_env_var(var_name: &str) -> Option<String> {
-    env::vars().find_map(|(key, value)| {
-        if key == var_name {
-            Some(value)
-        } else {
-            None
-        }
-    })
+    env::vars().find_map(
```
Comment (Chris00) [inline build.rs]: Why this reformatting?

File: build.rs [hunk 5]
```diff
@@ -69,8 +70,7 @@ fn build_vendored_sundials() -> (Option<String>, Option<String>, &'static str) {
     (lib_loc, inc_dir, library_type)
 }
 
-fn generate_bindings(inc_dir: &Option<String>)
-                     -> Result<Bindings, BindgenError> {
+fn generate_bindings(inc_dir: &Option<String>) -> Result<Bindings, BindgenError> {
```
Comment (Chris00) [inline build.rs]: Could you please refrain from modifying code that you do not change?

File: build.rs [hunk 6]
```diff
@@ -175,7 +230,15 @@ fn main() {
     }
 
     link! {"arkode", "cvode", "cvodes", "ida", "idas", "kinsol",
-           "nvecopenmp", "nvecpthreads"}
```
Comment (Chris00) [inline build.rs:178]: Why modify this line?

File: build.rs (hunk 4 above)
Comment (jlogan03) [reply build.rs]: rustfmt

File: Cargo.toml (hunk 3 above)
Comment (jlogan03) [reply Cargo.toml]: removed

File: .github/workflows/rust.yml (hunk 2 above)
Comment (jlogan03) [reply .github/workflows/rust.yml]: autoformatting from Prettier - reverted to the old version

File: build.rs (hunk 5 above)
Comment (jlogan03) [reply build.rs]: That change was consistent with rustfmt, but has been reverted

File: build.rs (hunk 6 above)
Comment (jlogan03) [reply build.rs:178]: reverted

File: build.rs (hunk 4 above)
Comment (jlogan03) [reply build.rs]: rolled that back as well

**Comment by jlogan03** [comment]
Rolled rustfmt formatting back and rebased to reduce commit debris

**Comment by Chris00** [comment]
As you may have noticed, I updated the sundials version to the latest (and compile the vendored version if the one present on the system is not the latest).

**Pull request closed without merging**